256 and RGB color modes

aremmell opened the pull request:
Creating pull request, but don't merge yet. I just want Travis to run it and make sure it's kosher.

aremmell commented:
@johnsonjh You want to take a look at these changes when you get a chance? I'm stuck on where to put the color mode. I stuck it where it is now because in the global config, it's even messier. I was thinking maybe it could just be part of the sir_textstyle struct I'm already holding on to, but that's per-level, and that'd be weird.

I just dunno.

aremmell commented on sirtextstyle.c:
```diff
@@ -117,187 +131,94 @@ bool _sir_resettextstyles(void) {
+                SIR_ESC, style->attr, _sir_getansifgcmd(style->fg), style->fg,
+                _sir_getansibgcmd(style->bg), style->bg);
         }
+        case SIRCM_RGB: {
```
This is bothering me. that buffer is 39 bytes, and those are a lot of integers. I think instead of making it the number of bytes it should be in theory, it should be the maximum that could expand to if every integer were maxed out

aremmell replied:
I don't trust snprintf to do the right thing. Best case scenario the log is jumbled up; worst–buffer overrun

aremmell replied:
I think what I should do is make `get*fromcolor` explicitly return `uint8_t`, and use `PRId8` here, so I can rest easy

aremmell commented on sirtypes.h:
```diff
@@ -77,49 +77,303 @@ typedef enum {
+    SIRTC_252      = 252,
+    SIRTC_253      = 253,
+    SIRTC_254      = 254,
+    SIRTC_255      = 255,
```
And this is kind of ridiculous, too. I don't think there needs to be an enum entry for every color from 17 to 255. It should just have a comment saying use the decimal form. We only need the last two.

aremmell commented:
> @johnsonjh You want to take a look at these changes when you get a chance? I'm stuck on where to put the color mode. I stuck it where it is now because in the global config, it's even messier. I was thinking maybe it could just be part of the sir_textstyle struct I'm already holding on to, but that's per-level, and that'd be weird.
> 
> I just dunno.

Figured it out. Stashed it with the level-to-style map, using the same mutex. This way there's only one lock. Also had to change mutexes to reentrant; I was getting deadlocks locking a mutex I already owned.

johnsonjh commented:
Sorry for the delay, got sidetracked with OpenBSD for a while too.  I'll probably get that in tonight and also check this out, but gonna grab a bite to eat.

aremmell commented:
> Sorry for the delay, got sidetracked with OpenBSD for a while too. I'll probably get that in tonight and also check this out, but gonna grab a bite to eat.



> Sorry for the delay, got sidetracked with OpenBSD for a while too. I'll probably get that in tonight and also check this out, but gonna grab a bite to eat.

No worries. Will you pay close attention to this one? It's a disruptive change, and my eyeballs are the ones who did it, so they won't notice anything is wrong.

johnsonjh commented:
> No worries. Will you pay close attention to this one? It's a disruptive change, and my eyeballs are the ones who did it, so they won't notice anything is wrong.

Yeah, I'll make sure it gets run through every compiler and system on the chart when you're ready.

aremmell commented:
> > No worries. Will you pay close attention to this one? It's a disruptive change, and my eyeballs are the ones who did it, so they won't notice anything is wrong.
> 
> Yeah, I'll make sure it gets run through every compiler and system on the chart when you're ready.

Ready. The stuff I just submitted isn't functional in nature, just formatting.

aremmell commented:
> > > No worries. Will you pay close attention to this one? It's a disruptive change, and my eyeballs are the ones who did it, so they won't notice anything is wrong.
> > 
> > 
> > Yeah, I'll make sure it gets run through every compiler and system on the chart when you're ready.
> 
> Ready. The stuff I just submitted isn't functional in nature, just formatting.

Welp, MSVC doesn't like the invalid enum being negative, so I guess I'm changing that. get a bunch of warnings about signed/unsigned mismatch, and found some dead code in _sir_formatstyle.

aremmell commented:
GTG on Windows/VS.

johnsonjh commented:
Busy day, I'll start testing in an hour or two!

johnsonjh commented:
@aremmell I'd recommend installing "advpng" - it's part of the advancecomp package. If you have it, "make docs" will automatically use it.  I just noticed a couple of the PNGs grew larger by a small amount and assume that is why.

johnsonjh commented:
@aremmell 

Some warnings:

```sh

sirhelpers.c: In function ‘_sir_validtextcolor’:
sirhelpers.c:194:54: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
  194 |             valid = SIRTC_DEFAULT == color || (color >= 0 && color <= 255);
      |                                                      ^~
sirhelpers.c:203:53: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
  203 |             valid = (SIRTC_DEFAULT == red   || (red >= 0 && red <= 255)) &&
      |                                                     ^~
sirhelpers.c:204:55: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
  204 |                     (SIRTC_DEFAULT == green || (green >= 0 && green <= 255)) &&
      |                                                       ^~
sirhelpers.c:205:54: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
  205 |                     (SIRTC_DEFAULT == blue  || (blue >= 0 && blue <= 255));
      |                                                      ^~
```

Oracle C is reporting:

`"sirtextstyle.c", line 152: warning: statement not reached`

this is a stray break below a return.  If you do want to put a break statement in there to close out each case statement (even though they don't get reached), adding ` /*NOTREACHED*/ /* unreachable */` (with that exact spacing) above each unreached break does suppress the warning for all the compilers we support.

I'd just remove that break though, but it's up to you. :)

johnsonjh commented:
@aremmell 

NVIDIA HPC SDK C reports the following new warnings:

```sh


"tests/tests.c", line 596: warning: integer conversion resulted in a change of sign [integer_sign_change]
      pass &= !sir_settextstyle(SIRL_ALERT, SIRTA_NORMAL, -1, -1);
                                                              ^
"tests/tests.c", line 1597: warning: integer conversion resulted in a change of sign [integer_sign_change]
          sir_textcolor fg = SIRTC_INVALID;
                             ^
"tests/tests.c", line 1598: warning: integer conversion resulted in a change of sign [integer_sign_change]
          sir_textcolor bg = SIRTC_INVALID;
```

johnsonjh commented:
@aremmell There is also a bug where file.exists is not created unless "make tests" is explicitly done.  That's my fault and I'll push a fix here.

johnsonjh commented on Makefile:
```diff
@@ -133,8 +133,8 @@ $(BINDIR)/file.exists:
-tests: $(OUT_TESTS) $(BINDIR)/file.exists
-$(OUT_TESTS): $(OUT_STATIC) $(OBJ_TESTS)
+tests: $(OUT_TESTS)
+$(OUT_TESTS): $(OUT_STATIC) $(OBJ_TESTS) $(BINDIR)/file.exists
```
See comments in thread.

johnsonjh replied:
Fixed this one myself, I'll leave the others to you.

johnsonjh reviewed (changes requested):
See comments above.

aremmell commented:
> @aremmell
> 
> Some new warnings from GCC 13 and NVIDIA C:
> 
> ```shell
> sirhelpers.c: In function ‘_sir_validtextcolor’:
> sirhelpers.c:194:54: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
>   194 |             valid = SIRTC_DEFAULT == color || (color >= 0 && color <= 255);
>       |                                                      ^~
> sirhelpers.c:203:53: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
>   203 |             valid = (SIRTC_DEFAULT == red   || (red >= 0 && red <= 255)) &&
>       |                                                     ^~
> sirhelpers.c:204:55: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
>   204 |                     (SIRTC_DEFAULT == green || (green >= 0 && green <= 255)) &&
>       |                                                       ^~
> sirhelpers.c:205:54: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
>   205 |                     (SIRTC_DEFAULT == blue  || (blue >= 0 && blue <= 255));
>       |                                                      ^~
> ```
> 
> Oracle C is also reporting:
> 
> ```shell
> "sirtextstyle.c", line 152: warning: statement not reached
> ```
> 
> ... this is from a stray break below a return. If you do want to put a break statement in there to close out each case statement (even though they don't get reached), adding:
> 
> ```c
>  /*NOTREACHED*/ /* unreachable */
> ```
> 
> ... (with that exact spacing) above each unreached break does suppress the warning for all the compilers we support.
> 
> I'd just remove that break though, but it's up to you. :)

I swear to Christ I removed that break last night. I must really be losing it now. I'm 100% sure I did. I must've on Windows and then forgotten to commit it.

aremmell commented:
> > @aremmell
> > Some new warnings from GCC 13 and NVIDIA C:
> > ```shell
> > sirhelpers.c: In function ‘_sir_validtextcolor’:
> > sirhelpers.c:194:54: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
> >   194 |             valid = SIRTC_DEFAULT == color || (color >= 0 && color <= 255);
> >       |                                                      ^~
> > sirhelpers.c:203:53: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
> >   203 |             valid = (SIRTC_DEFAULT == red   || (red >= 0 && red <= 255)) &&
> >       |                                                     ^~
> > sirhelpers.c:204:55: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
> >   204 |                     (SIRTC_DEFAULT == green || (green >= 0 && green <= 255)) &&
> >       |                                                       ^~
> > sirhelpers.c:205:54: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
> >   205 |                     (SIRTC_DEFAULT == blue  || (blue >= 0 && blue <= 255));
> >       |                                                      ^~
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Oracle C is also reporting:
> > ```shell
> > "sirtextstyle.c", line 152: warning: statement not reached
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ... this is from a stray break below a return. If you do want to put a break statement in there to close out each case statement (even though they don't get reached), adding:
> > ```c
> >  /*NOTREACHED*/ /* unreachable */
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ... (with that exact spacing) above each unreached break does suppress the warning for all the compilers we support.
> > I'd just remove that break though, but it's up to you. :)
> 
> I swear to Christ I removed that break last night. I must really be losing it now. I'm 100% sure I did. I must've on Windows and then forgotten to commit it.
> 
> On a technicality, since those values came from a uint8_t cast, yeah, they can never be negative, but the `sir_textcolor` type itself can. I suppose with the uint8_t cast, I don't have to check anything–any value in a byte is a valid RGB value.

You also just exposed another bug I just added last night; I set the `SIRTC_DEFAULT` to 256, which can't fit in uint8_t

aremmell commented:
Wait, don't review yet; have to fix the 256 problem.

aremmell commented:
Getting this for `make clean tests` now:

```
touch: ./build/bin/file.exists: No such file or directory
make: *** [build/bin/file.exists] Error 1
```

aremmell commented:
> @johnsonjh Getting this for `make clean tests` now:
> 
> ```
> touch: ./build/bin/file.exists: No such file or directory
> make: *** [build/bin/file.exists] Error 1
> ```

Looks like `build/bin` isn't there.

johnsonjh commented:
I'll fix it, sorry.

johnsonjh commented:
> Looks like `build/bin` isn't there.

Fixed now, and in the PR version.

aremmell commented:
> > Looks like `build/bin` isn't there.
> 
> Fixed now, and in the PR version.

Gracias. Now I have to figure out how I broke both 256 and RGB modes.

aremmell commented:
@johnsonjh K, let your demon compilers at it. I'm sure there's more fun in store.

johnsonjh commented:
> @johnsonjh K, let your demon compilers at it. I'm sure there's more fun in store.

GCC 13 reporting:
```sh
sirhelpers.c: In function ‘_sir_validtextcolor’:
sirhelpers.c:194:54: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
  194 |             valid = SIRTC_DEFAULT == color || (color >= 0 && color <= 255);
      |                                                      ^~
```

aremmell commented:
Ok, go. I forgot that I typedef'd sir_textcolor and it's not an enum. Too many late night changes.

johnsonjh commented:
> Ok, go. I forgot that I typedef'd sir_textcolor and it's not an enum. Too many late night changes.

Fixed that.  Hilariously, Oracle's Linter crashes with the most useless error ever (which isn't your fault):
```sh
"sirinternal.c", line 1234: internal compiler error: Wasted space
```

aremmell commented:
> > Ok, go. I forgot that I typedef'd sir_textcolor and it's not an enum. Too many late night changes.
> 
> Fixed that. Hilariously, Oracle's Linter gets crashes with the most useless error ever (which isn't your fault):
> 
> ```shell
> "sirinternal.c", line 1234: internal compiler error: Wasted space
> ```

Wow, never seen anything like it.

johnsonjh commented:
> > Fixed that. Hilariously, Oracle's Linter gets crashes with the most useless error ever (which isn't your fault):
> 
> > ```shell
> > "sirinternal.c", line 1234: internal compiler error: Wasted space
> > ```
> 
> Wow, never seen anything like it.

Not worried about it, the compiler is fine with it, it's just using the linter.  All other files pass.

johnsonjh commented:
@aremmell If you would, add `*.ln` to `.gitignore`, those are files output by Oracle Lint (and sometimes get leftover.)

johnsonjh commented:
Looks like I'm getting the right behavior and clean compilation with all the offbeat or extra picky stuff: Intel C, Oracle C, Oracle Lint, NVIDIA C, IBM XL C and Open XL, etc. Clang Analyzer and Cppcheck are good too.

I didn't look at every line of code, but as long as SIRCM_16 is going to remain the default mode, then this is good!

Related but for another PR:  I think I want to make the example application do some of the tricky stuff I mentioned earlier (like looking for the TERM variable being set to anything and the COLORTERM variable also being set to either24-bit" or "truecolor" and then outputting some additional RGB-mode lines).

aremmell commented:
> Looks like I'm getting the right behavior and clean compilation with all the offbeat or extra picky stuff: Intel C, Oracle C, Oracle Lint, NVIDIA C, IBM XL C and Open XL, etc. Clang Analyzer and Cppcheck are good too.
> 
> I didn't look at every line of code, but as long as SIRCM_16 is going to remain the default mode, then this is good!
> 
> Related but for another PR: I think I want to make the example application do some of the tricky stuff I mentioned earlier (like looking for the TERM variable being set to anything and the COLORTERM variable also being set to either "24-bit" or "truecolor" and then switching modes and outputting some additional RGB-mode lines).

That would be cool. We should definitely demonstrate the capability now that it's there, but we have to get people to find libsir and be convinced to use it.

It's not getting any traction. I fear it's too late and we'd be better off doing it in Rust or freakin Python.

Speaking of that, if Python can do C bindings, I say that's what we do next–turn libsir into a Python module. It's better than the one everyone uses (look at my `namecheap-ddns` repo; it uses the standard logger in Python)

johnsonjh commented:
> That would be cool. We should definitely demonstrate the capability now that it's there, but we have to get people to find libsir and be convinced to use it.
> 
> It's not getting any traction. I fear it's too late and we'd be better off doing it in Rust or freakin Python.
> 
> Speaking of that, if Python can do C bindings, I say that's what we do next–turn libsir into a Python module. It's better than the one everyone uses (look at my `namecheap-ddns` repo; it uses the standard logger in Python)

I Do Not Like Python.  I usually use REXX or Perl for scripting languages, but those are very hip these days either.  So I couldn't help much with Python, but it's probably a good idea to make a module.  For Perl, there are plenty of other options and people likely wouldn't use it for various legacy reasons, but I might look into what it would take to make a module.

I'll volunteer for Dlang bindings too, but it'll be *after* I switch DPS8M over to libsir.

aremmell commented:
> > That would be cool. We should definitely demonstrate the capability now that it's there, but we have to get people to find libsir and be convinced to use it.
> > It's not getting any traction. I fear it's too late and we'd be better off doing it in Rust or freakin Python.
> > Speaking of that, if Python can do C bindings, I say that's what we do next–turn libsir into a Python module. It's better than the one everyone uses (look at my `namecheap-ddns` repo; it uses the standard logger in Python)
> 
> I Do Not Like Python. I usually use REXX or Perl for scripting languages, but those are very hip these days either. So I couldn't help much with Python, but it's probably a good idea to make a module. For Perl, there are plenty of other options and people likely wouldn't use it for various legacy reasons, but I might look into what it would take to make a module.
> 
> I'll volunteer for Dlang bindings too, but it'll be _after_ I switch DPS8M over to libsir.

Well, once you do that, at least we'll have ONE project to claim that has it in use lol. I stuck it in some internal work code that 20 people will ever see in the future of planet earth, and used it in an IoT project that got shelved after I spent years making it perfect (that's actually what prompted the rewrite in 2018).

I put it in my version stamper, `mkverobj`, but seems like nobody needs that either. Time to learn Rust lol.

aremmell commented:
Is this good to merge? It still says changes requested, but I can't find where they are.

johnsonjh commented:
Good to go by me.

johnsonjh commented:
> Well, once you do that, at least we'll have ONE project to claim that has it in use lol. I stuck it in some internal work code that 20 people will ever see in the future of planet earth, and used it in an IoT project that got shelved after I spent years making it perfect (that's actually what prompted the rewrite in 2018).

The most "popular"( by GitHub-as-social-media standards) thing I have out there is [OpenVi](https://github.com/johnsonjh/OpenVi).  Which is hilarious to me because it's one many editors that I have on there!  And it wasn't even done for me, it was someone on Discord complaining that nvi2 is harder to build than it should be (and it uses the system regex libraries so it's not 100% consistent across platforms.)  They also had a bunch of arbitrary wishes (single file, no dependencies, etc.) So I did the ever-popular thing of taking OpenBSD's vi, regex, and BDB2 code, and making it work everywhere else.  Somehow people really like it, and it has actual hundreds and hundreds of users and it's even in some distros now.  Also got a few bug fixes into nvi2 and back to OpenBSD, which is nice.   But you'd think a minimal portable `vi` wouldn't be my most popular project.  You never know what people will like or use I guess.

I looked at a lot of different log things, but none of them met all the requirements I wanted/needed, especially portability and not needing a ton of weird dependencies.  And not being Google glog.  I probably should have reached out to you that I was considering using it way sooner!  But, I saw you working on it again, so I figured I might as well get my changes up to you, so I'm not stuck maintaining my own fork. :)

The other choice that would have been acceptable was [zlog](https://github.com/HardySimpson/zlog), but decided against it for several reasons.

johnsonjh commented:
Oh, moving DPS8M codebase from C99 to C11 was in part so we could use this library.  Nobody complained about that, as we'd been C99 since forever, which was often a pain for some things - like we have 5 different atomics modes, etc.

aremmell commented:
Bless you for even being able to use vi. If it weren't for nano I couldn't operate a *nix machine. I know how to :wq to get the fuck out, and that's it.
